### PR TITLE
Jpeg decoder optimization

### PR DESCRIPTION
**Pull request by br3aker**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
Finally I've managed to put this all together - part of the decoder optimization, color conversion & IDCT optimization to be precise. 

**Complete list of what was changed**
1. New IDCT method & fused zigzag/transpose steps from #1778
2. Color conversion refactoring
2.1 Removed obsolete bits of code from color converters (mainly Vector4 code from grayscale -> rgb scalar implementation)
2.2 `Vector8` grayscale implementation
2.3 Explicit Avx converters no longer do size checks and double virtual calls - JPEG format guarantees that pixel stride would be divisble by 8 so we don't need any size checks for avx/sse converters
3. Added somewhat 'baseline' jpeg decoder benchmark

**Benchmarks**
Master
|                              Method |      Mean |     Error |    StdDev |
|------------------------------------ |----------:|----------:|----------:|
|        'Baseline Interleaved 4:4:4' | 12.710 ms | 0.1120 ms | 0.0990 ms |
|        'Baseline 4:2:0 Interleaved' |  8.855 ms | 0.1447 ms | 0.1353 ms |
|        'Baseline 4:0:0 (grayscale)' |  1.660 ms | 0.0106 ms | 0.0088 ms |
| 'Progressive 4:2:0 Non-Interleaved' | 14.138 ms | 0.2797 ms | 0.3330 ms |

PR
|                              Method |      Mean |     Error |    StdDev |
|------------------------------------ |----------:|----------:|----------:|
|        'Baseline 4:4:4 Interleaved' | 10.822 ms | 0.0521 ms | 0.0487 ms |
|        'Baseline 4:2:0 Interleaved' |  8.386 ms | 0.0301 ms | 0.0282 ms |
|        'Baseline 4:0:0 (grayscale)' |  1.482 ms | 0.0052 ms | 0.0043 ms |
| 'Progressive 4:2:0 Non-Interleaved' | 13.005 ms | 0.0406 ms | 0.0380 ms |

### CODE REVIEW

**Comment by br3aker** [comment]
Next step - huffman decoder optimization, didn't want to smash every single change in this single PR cause there's a lot of things to optimize.

File: src/ImageSharp/Formats/Jpeg/Components/Decoder/ColorConverters/JpegColorConverter.FromYCbCrScalar.cs [hunk 1]
```diff
@@ -1,16 +1,21 @@
...
 
 namespace SixLabors.ImageSharp.Formats.Jpeg.Components.Decoder.ColorConverters
 {
     internal abstract partial class JpegColorConverter
     {
-        internal sealed class FromYCbCrBasic : BasicJpegColorConverter
+        internal sealed class FromYCbCrScalar : ScalarJpegColorConverter
         {
-            public FromYCbCrBasic(int precision)
+            // TODO: comments, derived from ITU-T Rec. T.871
```
[the remark this answers is not in the document]
Comment (br3aker) [reply src/ImageSharp/Formats/Jpeg/Components/Decoder/ColorConverters/JpegColorConverter.FromYCbCrScalar.cs:12]: Oh yeah, completely forgot about this, will add comments, thanks for pointing it out!

**Comment by br3aker** [comment]
There's a bug in one of the code paths which I'll fix later but that 'process crashed' from remote executor is strange.

File: src/ImageSharp/Formats/Jpeg/Components/Decoder/ColorConverters/JpegColorConverter.FromYCbCrVector8.cs [hunk 2]
```diff
@@ -1,18 +1,16 @@
...
 using System.Runtime.InteropServices;
-using SixLabors.ImageSharp.Tuples;
 
 // ReSharper disable ImpureMethodCallOnReadonlyValueField
 namespace SixLabors.ImageSharp.Formats.Jpeg.Components.Decoder.ColorConverters
 {
     internal abstract partial class JpegColorConverter
     {
-        internal sealed class FromYCbCrVector8 : Vector8JpegColorConverter
+        internal sealed class FromYCbCrVector8 : VectorizedJpegColorConverter
```
[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/ImageSharp/Formats/Jpeg/Components/Decoder/ColorConverters/JpegColorConverter.FromYCbCrVector8.cs:13]: > But it degrades Avx converters performance for some reason

Oh wow, that's crazy. Ok, I'm fine with a separate PR.

**Comment by antonfirsov** [comment]
> 'process crashed' from remote executor is strange.

Might be some intermittent issue, rerunning the tests.

> Explicit Avx converters no longer do size checks and double virtual calls

Can you help me by pointing to the particular code change for this?

**Comment by br3aker** [comment]
> Can you help me by pointing to the particular code change for this?

All Avx converters are derived from this class on current master:

https://github.com/SixLabors/ImageSharp/blob/d8d26162ae97815749e0aabd3a5d879e793ed0f9/src/ImageSharp/Formats/Jpeg/Components/Decoder/ColorConverters/JpegColorConverter.VectorizedJpegColorConverter.cs#L21-L40

While this is fine for something larger than 4 or 8 floats per vector - avx512 for example, we need to do these checks and do virtual calls. Avx specific implementations do not need to do any checks as jpeg provides color strides with lengths divisible by 8 due to 8x8 blocks. In this PR all avx converters are derived from base `JpegColorConverter` class without any size checks or extra virtual calls. You can see it in any `FromColorSpaceAvx.cs` code.

**Comment by br3aker** [comment]
Found a tremendous bug introduced who knows when (it was there before my decoder memory optimization):
https://github.com/SixLabors/ImageSharp/blob/7d74c4c95a8fb9c2df614c9dab708c4cf4b9181c/src/ImageSharp/Formats/Jpeg/Components/Decoder/SpectralConverter%7BTPixel%7D.cs#L71-L72

While first component usualy is the biggest one in terms of sampling, it's not always true, we even have test image as an example: [jpeg422.jpg](https://github.com/SixLabors/ImageSharp/blob/master/tests/Images/Input/Jpg/baseline/jpeg422.jpg).

```
  Number of Img components = 3
    Component[1]: ID=0x01, Samp Fac=0x11 (Subsamp 1 x 2), Quant Tbl Sel=0x00 (Lum: Y)
    Component[2]: ID=0x02, Samp Fac=0x12 (Subsamp 1 x 1), Quant Tbl Sel=0x01 (Chrom: Cb)
    Component[3]: ID=0x03, Samp Fac=0x12 (Subsamp 1 x 1), Quant Tbl Sel=0x01 (Chrom: Cr)
```

Certain test is now failing because I've removed a strange height if-check, this bug causes exceptions at least for jpeg422.jpg image test, maybe it'll even fix remote crash. That if-check did suppress actual error with sampling factors.

**Comment by JimBobSquarePants** [comment]
@br3aker Ooft. Well spotted!

**Review by antonfirsov — COMMENTED** [review]
Looks good mostly, but we should resolve the issue pointed out in https://github.com/SixLabors/ImageSharp/pull/1816#issuecomment-965916098. 

I also see the seemingly unrelated `RemoteExecutor_FailingRemoteTestShouldFailLocalTest` failing frequently in #1730, where I definitely have some lurking memory corruption bugs, I wonder what will is the case in this PR.

Another thing we must resolve, (and CC @JimBobSquarePants on that) is to agree on a story for platforms without SIMD. I'm afraid this PR may regress them. We may decide to ignore these concerns, but I'm not sure if it would be wise decision. I expect growing ImageSharp usage with Blazor for example. With [AOT](https://blog.elmah.io/ahead-of-time-compilation-for-blazor-wasm/), it's probably not as hopeless as reported in #1544.

**Comment by br3aker** [comment]
I'm sorry, was very busy this week, will try to finish PR this weekend.

> Another thing we must resolve, (and CC @JimBobSquarePants on that) is to agree on a story for platforms without SIMD. I'm afraid this PR may regress them.

Apart from removing SSE color conversion path for YCbCr -> Rgb and `Vector4` stuff for scalar code, what's else might regress performance?
If you're talking about extra transpose step in IDCT - latest release had 2 transpose calls in scalar implementation, I've removed it in one of my PRs. This PR introduces 1 transpose per IDCT so it's still a win compared to the latest release. We still can do IDCT without trasposes at all on non-simd platforms but that would introduce a lot of ugly code imo.

**Comment by antonfirsov** [comment]
> I'm sorry, was very busy this week

No worries, same here :)

> what else might regress performance?

Sorry, this wasn't clear. I only meant the `Vector4` stuff mentioned in https://github.com/SixLabors/ImageSharp/pull/1816#discussion_r748726264.

File: src/ImageSharp/Formats/Jpeg/Components/Block8x8.cs [hunk 3]
```diff
@@ -337,6 +337,64 @@ public nint GetLastNonZeroIndex()
             }
         }
 
+        /// <summary>
+        /// Scalar inplace transpose implementation for <see cref="TransposeInplace"/>
+        /// </summary>
+        [MethodImpl(InliningOptions.ShortMethod)]
+        public void TransposeInplace()
```
[the remark this answers is not in the document]
Comment (br3aker) [reply src/ImageSharp/Formats/Jpeg/Components/Block8x8.cs:344]: Yeah, will move it to the test code as extension method;.

**Comment by JimBobSquarePants** [comment]
>Another thing we must resolve, (and CC @JimBobSquarePants on that) is to agree on a story for platforms without SIMD. I'm afraid this PR may regress them. We may decide to ignore these concerns, but I'm not sure if it would be wise decision. I expect growing ImageSharp usage with Blazor for example. With AOT, it's probably not as hopeless as reported in #1544.

What's the SIMD story with WebAssembly now? I know there's been some progress there. We should always avoid any non SIMD support regression where possible.

**Comment by br3aker** [comment]
I still don't understand what will cause regression here :/

Current non-simd color converters use `float`s, not `Vector4`s. This PR does the same. Proposal by Anton will improve current scalar performance but it has nothing to do with this PR.

**Comment by JimBobSquarePants** [comment]
>I still don't understand what will cause regression here :/

Actually I can't see one either (from a quick) read. The scalar stuff seems to be lift and shift

File: src/ImageSharp/Formats/Jpeg/Components/Decoder/ColorConverters/JpegColorConverter.FromCmykVector8.cs [hunk 4]
```diff
@@ -40,14 +37,14 @@ protected override void ConvertCoreVectorizedInplace(in ComponentValues values)
                     ref Vector<float> y = ref Unsafe.Add(ref yBase, i);
                     Vector<float> k = Unsafe.Add(ref kBase, i) * scale;
 
-                    c = (c * k) * scale;
-                    m = (m * k) * scale;
-                    y = (y * k) * scale;
+                    c = c * k * scale;
+                    m = m * k * scale;
+                    y = y * k * scale;
```
Comment (gfoidl) [inline src/ImageSharp/Formats/Jpeg/Components/Decoder/ColorConverters/JpegColorConverter.FromCmykVector8.cs:42]:
```suggestion
                    k = k * scale;
                    c = c * k;
                    m = m * k;
                    y = y * k;
```
The JIT [doesn't CSE this](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgDkBXfGKASzFwBuGvSasAShwB2GXtxYBhCPgAOvADY8AyjwBu/GEJoiAzE1IMFDAN40G9pmYBm6iNgwMAstgQA1bOocMDYMAOYwGIIMKny67sG4EVEAvnYOafbEZsQoXuQAFLBODL4wYBjQADwubhgAfAxgAEJ4MGgMRSVlFVDVru4N+C2J7Z2l5VU1AwwAnsNtHTDF4z19tQ0A1vPtUrwyDFIAlAwZNqcODHFQDLhgAcEAvAcwAO5dE71T9fnkDAD0DAwAAteLgWN4/AEgodhNQLhdzg4nNAGPldvteAwnnQopjKgdcQBqQmHRH2Wxw+FUsbdSb9eqNLGLYoAVSkuGwThgLAAggATPmFJaNbYMXgwsnwmkfNbTfBMzpsjlc3kCoXFIatdri2FU6nClZ09azBXCpWc7n8wWdOZasUSyl6hyGz70zZM80qq3qhhbO3ihgAKhud00sMlFzATKjwY2QZD911Toc8qe8tj8duiYjDhmTLzGeDWbDktSjoYZYR5ayTFynlIPpdsoZzTt0tWX0GovbRumtpGzPeHbdvtF6I8RxO5YpTquCc0TKkryHve+vwBwNB4J8/kCMAdTslyOuaL2Hkx2NxDHxUiJJMlM+T9h7ruNUaeivZFtV1uFrYHOo5s+Bq0q+cqmqyX5emqnSagBB5PoOTadiaH5mlBlowcK/YLIB5ZOshI5xk8nqYb+xR+vBmahjASZPkBvpMnGRY0XRybvoysZsU6qYMOmvrcXqeZPAWAmlqcZbJEAA=) (?!). So some `mulps` instructions can be saved.

File: src/ImageSharp/Formats/Jpeg/Components/Decoder/ColorConverters/JpegColorConverter.VectorizedJpegColorConverter.cs [hunk 5]
```diff
@@ -1,27 +1,36 @@
...
-                this.vectorSize = vectorSize;
             }
 
+            protected sealed override bool IsAvailable => SimdUtils.HasVector8;
+
             public override void ConvertToRgbInplace(in ComponentValues values)
             {
                 int length = values.Component0.Length;
-                int remainder = values.Component0.Length % this.vectorSize;
+                int remainder = values.Component0.Length % 8;
```
Comment (gfoidl) [inline src/ImageSharp/Formats/Jpeg/Components/Decoder/ColorConverters/JpegColorConverter.VectorizedJpegColorConverter.cs:33]:
```suggestion
                int remainder = values.Component0.Length & 7;   // bit-hack for % 8
```

**Comment by br3aker** [comment]
I've (hopefully) fixed mentioned bug but CI seems to be stuck for some reason. I will psuh fixes for reviews a bit later.

**Comment by br3aker** [comment]
There's definitely some invalid memory-related code, problem is, I have no idea how to find it considering all jpeg tests pass...

**Comment by br3aker** [comment]
This is undetectable in a reasonable amount of time IMO, closing this so I can introduce each change from this PR as a separate PR so potential memory leack/out-of-bounds will be more manageable.

**Comment by antonfirsov** [comment]
@br3aker some ideas:

- Disable `RemoteExecutor_FailingRemoteTestShouldFailLocalTest` since it's probably noise, hiding the real issue
- Try merging your branch with [UniformUnmanagedMemoryPoolMemoryAllocator-02-MemoryGuards](https://github.com/SixLabors/ImageSharp/commits/af/UniformUnmanagedMemoryPoolMemoryAllocator-02-MemoryGuards). Although it has it's own issues, there is a trick to detect buffer overflows on `MemoryAllocator` buffer disposal: a329a3a8a27901377d19b617d26e341183a5942b, and the new strongly discontiguous memory patterns may also help revealing issues
- Try running tens of `dotnet test` runs in a loop (eg. overnight), collecting the failure logs:  https://github.com/SixLabors/ImageSharp/blob/af/UniformUnmanagedMemoryPoolMemoryAllocator-02-MemoryGuards/tests/ImageSharp.Tests/RunTestsInLoop.ps1

**Comment by antonfirsov** [comment]
Though redoing the changes step-by-step is a very wise plan!

**Pull request closed without merging**